File: frontend/src/api/volunteers.ts
import axios from "axios";

import type { Assignment, Volunteer, VolunteerCreate } from "../types";

const api = axios.create({
  baseURL: import.meta.env.VITE_API_URL || "http://localhost:8000",
});

type ApiErrorPayload = {
  detail?: string | { message?: string } | Array<{ msg?: string } | string>;
  message?: string;
};

function toApiError(error: unknown): Error {
  if (axios.isAxiosError<ApiErrorPayload>(error)) {
    const payload = error.response?.data;

    if (typeof payload === "string") {
      return new Error(payload);
    }

    if (typeof payload?.detail === "string") {
      return new Error(payload.detail);
    }

    if (Array.isArray(payload?.detail) && payload.detail.length > 0) {
      const first = payload.detail[0];
      if (typeof first === "string") {
        return new Error(first);
      }
      if (typeof first?.msg === "string") {
        return new Error(first.msg);
      }
    }

    if (
      typeof payload?.detail === "object" &&
      payload?.detail !== null &&
      "message" in payload.detail &&
      typeof payload.detail.message === "string"
    ) {
      return new Error(payload.detail.message);
    }

    if (typeof payload?.message === "string") {
      return new Error(payload.message);
    }

    return new Error(error.message || "Request failed");
  }

  if (error instanceof Error) {
    return new Error(error.message);
  }

  return new Error("Request failed");
}

export async function registerVolunteer(data: VolunteerCreate): Promise<Volunteer> {
  try {
    const response = await api.post<Volunteer>("/api/volunteers/register", data);
    return response.data;
  } catch (error) {
    throw toApiError(error);
  }
}

export async function getAvailableVolunteers(): Promise<Volunteer[]> {
  try {
    const response = await api.get<Volunteer[]>("/api/volunteers/", {
      params: { available_only: true },
    });
    return response.data;
  } catch (error) {
    throw toApiError(error);
  }
}

export async function updateAvailability(id: string, available: boolean): Promise<Volunteer> {
  try {
    const response = await api.patch<Volunteer>(`/api/volunteers/${id}/availability`, {
      availability: available,
    });
    return response.data;
  } catch (error) {
    throw toApiError(error);
  }
}

export async function getVolunteerTasks(id: string): Promise<Assignment[]> {
  try {
    const response = await api.get<Assignment[]>(`/api/volunteers/${id}/tasks`);
    return response.data;
  } catch (error) {
    throw toApiError(error);
  }
}

export async function getVolunteer(id: string): Promise<Volunteer> {
  try {
    const response = await api.get<Volunteer>(`/api/volunteers/${id}`);
    return response.data;
  } catch (error) {
    throw toApiError(error);
  }
}

File: backend/app/routers/volunteers.py
from uuid import UUID
from typing import List, Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import ValidationError

from app.db.supabase_client import supabase
from app.models.volunteer import (
    VolunteerCreate,
    VolunteerResponse,
    VolunteerAvailabilityUpdate,
)

router = APIRouter(prefix="/api/volunteers", tags=["volunteers"])


@router.post("/register", response_model=VolunteerResponse, status_code=201)
async def register_volunteer(volunteer: VolunteerCreate):
    """Register a new volunteer with their skills and location."""
    try:
        # Convert to JSON serializable dictionary (resolves UUIDs, Enums)
        data = volunteer.model_dump(mode="json")
        
        response = supabase.table("volunteers").insert(data).execute()
        
        if not response.data:
            raise HTTPException(status_code=400, detail="Failed to create volunteer")
            
        return response.data[0]
    except Exception as e:
        # Supabase API errors usually bubble up here
        raise HTTPException(status_code=400, detail=f"Database error: {str(e)}")


@router.get("/", response_model=List[VolunteerResponse])
async def list_volunteers(
    available_only: bool = Query(False, description="Filter for available volunteers only"),
    skill: Optional[str] = Query(None, description="Filter by a specific skill")
):
    """Retrieve a list of volunteers with optional filtering."""
    try:
        query = supabase.table("volunteers").select("*")
        
        if available_only:
            query = query.eq("availability", True)
            
        if skill:
            # Assumes Postgres array column 'skills'
            query = query.contains("skills", [skill])
            
        response = query.execute()
        return response.data
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Database error: {str(e)}")


@router.get("/{id}", response_model=VolunteerResponse)
async def get_volunteer(id: UUID):
    """Fetch details of a specific volunteer by UUID."""
    try:
        response = supabase.table("volunteers").select("*").eq("id", str(id)).execute()
        
        if not response.data:
            raise HTTPException(status_code=404, detail="Volunteer not found")
            
        return response.data[0]
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Database error: {str(e)}")


@router.patch("/{id}/availability", response_model=VolunteerResponse)
async def update_availability(id: UUID, update_data: VolunteerAvailabilityUpdate):
    """Toggle or update a volunteer's availability status."""
    try:
        response = (
            supabase.table("volunteers")
            .update({"availability": update_data.availability})
            .eq("id", str(id))
            .execute()
        )
        
        if not response.data:
            raise HTTPException(status_code=404, detail="Volunteer not found")
            
        return response.data[0]
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Database error: {str(e)}")


@router.get("/{id}/tasks")
async def get_volunteer_tasks(id: UUID):
    """Get all assignments for this volunteer joined with task details."""
    try:
        # Assuming an 'assignments' table joins volunteers with tasks:
        # assignments has volunteer_id and task_id. tasks(*) performs the join.
        response = (
            supabase.table("assignments")
            .select("*, tasks(*)")
            .eq("volunteer_id", str(id))
            .execute()
        )
        return response.data
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Database error: {str(e)}")


@router.delete("/{id}", response_model=VolunteerResponse)
async def delete_volunteer(id: UUID):
    """Soft delete a volunteer by marking them as unavailable."""
    try:
        # We soft-delete by setting availability to False
        response = (
            supabase.table("volunteers")
            .update({"availability": False})
            .eq("id", str(id))
            .execute()
        )
        
        if not response.data:
            raise HTTPException(status_code=404, detail="Volunteer not found")
            
        return response.data[0]
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Database error: {str(e)}")
